src/test/ui/block-result/issue-20862.stderr

```diff
@@ -12,8 +12,16 @@ LL |     |y| x + y
 error[E0618]: expected function, found `()`
   --> $DIR/issue-20862.rs:17:13
    |
-LL |     let x = foo(5)(2);
-   |             ^^^^^^^^^ not a function
+LL | / fn foo(x: i32) {
+LL | |     |y| x + y
+LL | | //~^ ERROR: mismatched types
+LL | | }
+   | |_- `foo` defined here returns `()`
+...
+LL |       let x = foo(5)(2);
+   |               ^^^^^^---
+   |               |
+   |               call expression requires function
```

Comment:
I wonder how muddled it'd be if you kept the "not a function" label pointing at the primary span. Even better, detecting chained calls would be nice to provide a custom message (follow up work that needs more thought put into it):

```
error[E0618]: expected function, found `()`
  --> $DIR/issue-20862.rs:17:13
   |
LL | / fn foo(x: i32) {
LL | |     |y| x + y
LL | | //~^ ERROR: mismatched types
LL | | }
   | |_- `foo` defined here returns `()`
...
LL |       let x = foo(5)(2);
   |               ^^^^^^---
   |               |
   |               call expression requires function
  = note: the highlighted code below resolves to `()`...
LL |       let x = foo(5)(2);
   |               ^^^^^^
  = note: ...which cannot be called with the arguments `(2)`
```